internal/restic/backend.go

```diff
@@ -32,10 +32,9 @@ type Backend interface {
 	// Stat returns information about the File identified by h.
 	Stat(ctx context.Context, h Handle) (FileInfo, error)
 
-	// List returns a channel that yields all names of files of type t in an
-	// arbitrary order. A goroutine is started for this, which is stopped when
-	// ctx is cancelled.
-	List(ctx context.Context, t FileType) <-chan string
+	// List runs fn for each file in the backend which has the type t. When an
+	// error occurs (or fn returns an error), List stops and returns it.
+	List(ctx context.Context, t FileType, fn func(FileInfo) error) error
```

Comment:
I would mention that `fn` is allowed to call Backend. This has implications for how backend implementations throttle concurrent remote calls. 

I would also explicitly say that `fn` will be called exactly once for each `FileInfo` during successful `List()` execution (and at most once in case of an error). This has implications for `BackendRetry` implementation.

Reply:
Hmm, I think we also need to explicitly state that `fn` will be called on the calling gorouting. Otherwise each client will need to deal with synchronization issues if they want to [collect results in a map](https://stackoverflow.com/questions/11063473/map-with-concurrent-access) for example.

Reply:
Ah, good points, thanks. The `List()` backend call isn't restricted (in terms of parallel connections) in any existing backend, only the other operations like `Load()` and `Save()`.

> Hmm, I think we also need to explicitly state that fn will be called on the calling gorouting

Why do you feel that way? Sure, we can add that, but In my experience, APIs for libraries rather state if something isn't run in the calling goroutine...

Reply:
If "on the same gorouting" is the normal/expected way of doing things in Go, then indeed no need to mention it explicitly. (sorry, I am new to Go, still learning my ways around).

Reply:
We can still add it for clarity, I was just wondering if maybe I got something wrong :)